Add context manager for conditional rewrites of torch.* to torch._refs.* calls

IvanYashchuk opened the pull request:
Adds a new context manager `TorchRefsNvfuserCapabilityMode` for conditional rewrite of `torch.*` calls to `torch._refs.*` based on whether the decomposition consisting of prims supports nvFuser execution or not.

A new optional argument for `TorchRefsMode` is added - `should_fallback_fn`, a callable that returns whether the original `torch.foo` or the replacement `torch._refs.foo` should be used.

IvanYashchuk committed:
Add the first version of the refs mode based on nvfuser+prim capability

IvanYashchuk committed:
Add test_nvfuser_capability_context

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Add should_fallback_fn to TorchRefsMode

IvanYashchuk committed:
Move the code for nested make_fx to proxy_tensor.py

IvanYashchuk committed:
Add test_isolated_graphmodule

IvanYashchuk committed:
orig_func is not used

IvanYashchuk committed:
Fix default should_fallback_fn

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -408,3 +408,49 @@ def wrap_fake(x):
+    # make_fx doesn't support kwargs, so we need to do this flattening
+    # and then unflatten the args before calling func
+    nargs = len(args)
+    flat_kwargs = list(kwargs.values())
```
Consider using pytree flatten/unflatten instead?

IvanYashchuk replied:
Done.

IvanYashchuk committed:
Raise an error if no proxy tensor is found in the args

IvanYashchuk committed:
Use pytree

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into nvfuser-context

IvanYashchuk committed:
Context manager should work for normal non-ProxyTensor usage

jjsjann123 commented on torch/_prims/context.py:
```diff
@@ -87,6 +91,9 @@ def __torch_function__(
         mapping = torch_to_refs_map()
```
This function seems to be only targetting torch.xxx.

In our dynamo workload, I think we are expecting an input GraphModule with `aten.ops`. Might want to expand this.

IvanYashchuk replied:
Right, it's only torch.xxx for now. One option is to use dynamo under this context manager and another is to expand the context manager with "aten_to_refs_map", let's handle the extension in a separate PR.

jjsjann123 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -410,3 +410,47 @@ def wrap_fake(x):
+    return next(proxy_tensors, None)
+
+
+def get_isolated_graphmodule(func, args, kwargs):
```
Would we want some special decomposition here to short-cut things like `nvfuser.var_mean`? With a proper `impl_nvfuser` this short-cut would work out well :tada:

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -410,3 +410,47 @@ def wrap_fake(x):
+    try:
+        for arg in all_args:
+            if isinstance(arg, ProxyTensor):
+                arg.proxy.tracer = new_tracer
```
Instead of mutating the argument, couldn't you pull out the inner `elem` and create a fresh ProxyTensor from that? Then you wouldn't need to worry about resetting the proxy afterwards.

IvanYashchuk replied:
Yes I can and I posted this example of creating a fresh ProxyTensor on Slack (https://pytorch.slack.com/archives/C03DP57R27M/p1658256451021309?thread_ts=1658256439.210389&cid=C03DP57R27M)

It requires creating a fresh torch.fx.proxy.Proxy which in turn requires creating a fresh Node (because nodes do not work with arbitrary graphs, it has to match). Just resetting the tracer seemed cleaner. Would you like me to change that and create a fresh ProxyTensor instead?

IvanYashchuk replied:
I changed it now to create a fresh ProxyTensor instead of mutating.

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -410,3 +410,47 @@ def wrap_fake(x):
+            if isinstance(arg, ProxyTensor):
+                arg.proxy.tracer = new_tracer
+
+        gm = make_fx(wrapped)(all_args)
```
This doesn't feel like it is enough. If there is an ambient proxy tensor mode active, it was never disabled and will still be interposing on invocations. But your test is passing. Do you have an explanation for why this is working as is?

ezyang replied:
It looks like the answer is, you're in a context where the proxy tensor mode is disabled by the time you called this

IvanYashchuk replied:
It works because all the calls are recorded on the different tracer which is created in this function.

ezyang replied:
I see... this feels so bad haha

IvanYashchuk replied:
Do you have a different understanding of the expected behavior? We have ProxyTensors as inputs to the function they all are expected to have the same tracer object attached (there's an assert for this somewhere), this tracer object is the place where the information about the calls gets stored and when we swap the tracer information about the calls is getting stored elsewhere.

ezyang replied:
One reason it feels bad is because the ProxyTensorMode itself *contains a tracer*, and that tracer is now inconsistent with the tracers on the proxy objects. In fact, this means that if you have factory functions inside the conditional rewrite, they will go to the wrong graph. So this is definitely wrong!

IvanYashchuk replied:
Yes, this is definitely wrong.
I added a test case with factory functions and two nested `get_isolated_graphmodule` calls. The test passes with:
```
mode = torch._C._get_torch_dispatch_mode()
with enable_torch_dispatch_mode(mode.inner, replace=mode):
    make_fx(...)
```
Then I added a test case with two nested `make_fx` and one `get_isolated_graphmodule` with factory functions. It also now passes with ExistStack of `enable_torch_dispatch_mode(mode.inner, replace=mode)` contexts.

ezyang commented on torch/_prims/context.py:
```diff
@@ -95,3 +102,28 @@ def __torch_function__(
+
+
+TorchRefsNvfuserCapabilityMode = functools.partial(
+    TorchRefsMode.push, should_fallback_fn=_is_func_unsupported_nvfuser
```
per @samdow's recent PR you don't need `.push` here

IvanYashchuk replied:
Thanks, I will change that once https://github.com/pytorch/pytorch/pull/78215 is actually merged.

samdow replied:
@IvanYashchuk It should work to change this now. That PR is just to clean up the API so that we insist on only having one way pushing a mode (and all code reflects that way)

ezyang commented:
1. Write a function that walks up the current mode stack, and looks for ProxyTensorMode
2. Make every ProxyTensor hold a reference to ProxyTensorMode (you can look at FakeTensor to see an example of how this is done)
3. Add a boolean flag on ProxyTensorMode that says whether or not tracing is enabled or not
4. Inside `proxy_call` and inside `ProxyTensorDispatchMode`, before doing operations on proxies/writing to the graph, check if tracing is enabled. If it is not, do NOT do the operation, instead replace the proxy with some invalid value (say None).
5. Before doing operations on Proxies, test that they are valid, otherwise raise an error
6. Make isolated graph disables all proxy tensor modes on the mode stack (using (1) to find the modes), and then runs make_fx with the arguments as is.

ezyang commented:
If reentrant proxy tracing isn't working (@zou3519 mentioned that it might not be at the moment), that needs to get fixed first too

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into nvfuser-context

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into nvfuser-context

IvanYashchuk committed:
Add nested isolated graphmodule calls and use of torch factory functions

IvanYashchuk committed:
Disable ambient proxy tensor mode inside get_isolated_graphmodule

IvanYashchuk committed:
Create new detached proxy tensors instead of mutating in try: finally:

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into nvfuser-context

IvanYashchuk committed:
Remove getting old_tracer

IvanYashchuk committed:
ufmt

IvanYashchuk committed:
Remove .push()

IvanYashchuk committed:
Remove print_tabular

IvanYashchuk committed:
Make nested make_fx and isolated tensor factory functions work

IvanYashchuk commented:
Nested ambient `ProxyTensorMode` is now handled and there's a test case with make_fx ∘ make_fx ∘ isolated_graphmodule. I removed the mutation of ProxyTensor arguments, instead, fresh proxies are created using a new tracer.

I couldn't make up a test case for 
> What if a inner tensor from the isolated graph mode escapes (e.g. by mutation)

Would it be alright to handle it separately in the future? Same question for making ProxyTensors to hold a reference to ProxyTensorMode and use that?

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -527,3 +531,51 @@ def wrap_fake_symbolic(x, sym_shape):
+            ),
+        )
+        if isinstance(a, ProxyTensor) else a for a in all_args
+    ]
```
Wasn't the plan here to add a flag to the proxy tensor mode to disable tracing, so that you wouldn't have to all of this detaching business? If I understand correctly, you're going to end up with a malformed FX graph becausethese nodes refer to nodes that aren't actually in the graph.

IvanYashchuk replied:
A new graph is being created here and new "placeholder" nodes are created on the graph with `graph.create_node(...)`, so the nodes are actually in the graph, maybe it would be less confusing if I wrote `graph.placeholder(a.proxy.node.target)`. The resulting graph is not malformed.

For normal (not tensor factory functions) the trace is recorded not with the `ProxyTorchDispatchMode`, but with `torch.fx.Proxy`'s tracer and its `__torch_function__` override. Therefore we need new Proxy objects here, detached from the outer tracer in `ProxyTensor.proxy.tracer`. That's an explicit approach. Now with disabled ambient ProxyTorchDispatchModes a new isolated tracer and proxies are created implicitly inside the `make_fx` call just with unwrapped `ProxyTensor.elem`, so I will change the code to use that.

ezyang commented:
This is improved but it doesn't seem like it has all of the stuff we talked about

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into nvfuser-context

IvanYashchuk committed:
Unwrap elem instead of creating new torch.fx.Proxy

IvanYashchuk commented:
> 1. Write a function that walks up the current mode stack, and looks for ProxyTensorMode

Done with a combination of `maybe_disable_proxy_tensor_mode()` and `while torch._C._get_torch_dispatch_mode() is not None`.

> 2. Make every ProxyTensor hold a reference to ProxyTensorMode (you can look at FakeTensor to see an example of how this is done)

Done in https://github.com/pytorch/pytorch/pull/82549.

> 6.Make isolated graph disables all proxy tensor modes on the mode stack (using (1) to find the modes)

Done:
```py
with contextlib.ExitStack() as stack:
    while torch._C._get_torch_dispatch_mode() is not None:
        stack.enter_context(maybe_disable_proxy_tensor_mode())
```

> then runs make_fx with the arguments as is

`make_fx` is now run with unwrapped `elem` of proxy tensors. I think steps (3), (4), (5) are not needed in this case?

In the current state `get_isolated_graphmodule` is a very simple function that unwraps given proxy tensors and runs `make_fx` with all outer ProxyTensorModes disabled.

Could you please provide a concrete test case you think the current state of the pull request is not covering? Because as I mentioned I couldn't make up a test case for

> What if a inner tensor from the isolated graph mode escapes (e.g. by mutation)

ezyang commented:
> Could you please provide a concrete test case you think the current state of the pull request is not covering? Because as I mentioned I couldn't make up a test case for

The reason unwrapping doesn't unconditionally work is because the proxy tensor may itself be embedded within another data structure that isn't tree mappable. The most common situation is a tensor subclass. For example, consider this patch:

```
diff --git a/test/test_proxy_tensor.py b/test/test_proxy_tensor.py
index 9ac4e0a470..9d1635ecfa 100644
--- a/test/test_proxy_tensor.py
+++ b/test/test_proxy_tensor.py
@@ -169,7 +169,10 @@ class TestProxyTensor(TestCase):
             self.assertTrue(is_any_sum(gm))
             return torch.sigmoid(x)
 
+        from torch.testing._internal.logging_tensor import LoggingTensor
+
         def f2(x):
+            x = LoggingTensor(x)
             gm = get_isolated_graphmodule(f1, (x,), {})
             self.assertFalse(is_any_sum(gm))
             self.assertTrue(is_any_sigmoid(gm))
```

The LoggingTensor prevents the unwrapping from happening on the inside, and so sigmoid shows up in the inner graph.

If you *really* don't want to handle this case, I suppose that is reasonable, because inside AOTAutograd the expectation is that all tensor subclasses have already been erased, so we should only be passing plain tensors through and this should never happen. But in that case, there ought to be asserts about the assumed preconditions; and I also don't think it is that complicated to make it work for this general case as well.

ezyang commented:
You also have another problem which is that maybe disable proxy tensor mode cannot "see" if there is a proxy mode behind another, more recently pushed mode. When I patch with:

```
diff --git a/test/test_proxy_tensor.py b/test/test_proxy_tensor.py
index 9ac4e0a470..4cd100f18f 100644
--- a/test/test_proxy_tensor.py
+++ b/test/test_proxy_tensor.py
@@ -169,8 +169,11 @@ class TestProxyTensor(TestCase):
             self.assertTrue(is_any_sum(gm))
             return torch.sigmoid(x)
 
+        from torch.testing._internal.logging_tensor import LoggingTensorMode
+
         def f2(x):
-            gm = get_isolated_graphmodule(f1, (x,), {})
+            with LoggingTensorMode():
+                gm = get_isolated_graphmodule(f1, (x,), {})
             self.assertFalse(is_any_sum(gm))
             self.assertTrue(is_any_sigmoid(gm))
             return torch.digamma(x)
```

the script seems to infinite loop.

IvanYashchuk committed:
Fix infinite loop and interaction with non-ProxyTensor modes

IvanYashchuk committed:
Revert maybe_disable_tensor_mode

IvanYashchuk committed:
Assert there are no Tensor subclasses passed to make_fx in get_isolated_graphmodule

IvanYashchuk committed:
Add a link to the comment about nested tensor subclass wrappers

IvanYashchuk commented:
Thanks a lot for the help here!
I've fixed the infinite loop problem: currently, all dispatch modes are disabled first and then pushed again skipping proxy tensor modes.

I've added an assert that unwrapped Tensor arguments should not wrap other Tensors for now. I'd like to make it work later in a separate PR.

> inside AOTAutograd the expectation is that all tensor subclasses have already been erased, so we should only be passing plain tensors through and this should never happen.

Is this erasion done by TorchDynamo?

IvanYashchuk committed:
Add a comment why push is used instead of restore

ezyang commented:
> Is this erasion done by TorchDynamo?

It's done by AOTAutograd

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -527,3 +528,55 @@ def wrap_fake_symbolic(x, sym_shape):
+        getattr(a, "elem", None) is None
+        for a in unwrapped_all_args
+        if isinstance(a, torch.Tensor)
+    ), "ProxyTensor is wrapped with another Tensor subclass"
```
I don't feel like this assert actually works haha

IvanYashchuk replied:
Maybe... :thinking: But the test passes!

The assumption is that a tensor subclass is actually a subclass of `torch.Tensor` and not a generic object created with `torch.Tensor._make_wrapper_subclass`. Is there a better way to test for subclasses?

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -527,3 +528,55 @@ def wrap_fake_symbolic(x, sym_shape):
+        for mode in reversed([m for m in modes if not isinstance(m, ProxyTorchDispatchMode)]):
+            # mode.restore() doesn't work because mode.inner might be ProxyTorchDispatchMode
+            # mode.push() is restricted to modes that don't take any arguments
+            stack.enter_context(mode.push())
```
This still seems super error prone. As you say here, this only works for modes that don't take any arguments. It would be much better if there was just a boolean toggle on the proxy dispatch mode you can use to turn it off without actually removing it from the stack.

IvanYashchuk replied:
I just pushed a change to use `mode.restore()` with modified `mode.inner` and `mode.ancestors`, is this too hacky?

ezyang replied:
I'm against anything that modifies mode.inner, we don't really know how to understand dynamically changing mode stack structure, it's really weird and hard to think about.

IvanYashchuk replied:
Instead of the current approach, you would like to see the same as done with `in_kernel_invocation_manager` and `FakeTensorMode.in_kernel_invocation = True/False`? But it also mutates the attribute locally, yes not the `.inner` but a different attribute, still similar.

ezyang replied:
Yes, that is my preference. The mutation in this case can be treated like a dynamically scoped variable with limited impact (as opposed to changing of inner, which totally changes the semantics of subsequent calls in the stack.)

IvanYashchuk replied:
Okay, do you think changing of inner should be disallowed programmatically?

IvanYashchuk replied:
In this situation, the semantics of modes is quite clear and asserted that outside the "with" block the modes are not changed. Isn't it a good thing to exercise various usage of `.inner` attribute? 
Is there an issue for 
https://github.com/pytorch/pytorch/blob/afafd16671b2d2f253cad7c3d453ad1fc44cccf5/torch/utils/_python_dispatch.py#L24-L25

IvanYashchuk committed:
Use restore() instead of push()

IvanYashchuk committed:
Add asserts that modes outside of ExitStack are not modified

IvanYashchuk commented:
Now, this PR is safer: asserts are added to prevent nested tensor subclasses, and other asserts are added to verify that all tensor modes before and after the context disabling proxy tensor modes are the same and not modified.

The last thing to do for this PR to be accepted is to change the approach of disabling proxy tensor modes: https://github.com/pytorch/pytorch/pull/81764#discussion_r934741520 (blocked on https://github.com/pytorch/pytorch/pull/82549 being merged first).

IvanYashchuk commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -527,3 +529,58 @@ def wrap_fake_symbolic(x, sym_shape):
+        assert torch._C._get_torch_dispatch_mode() is None
+
+        # Enable all torch dispatch modes except ProxyTorchDispatchMode
+        for mode in reversed([m for m in modes if not isinstance(m, ProxyTorchDispatchMode)]):
```
@samdow, what do you think of the approach here modifying `.inner` and `.ancestors` to rebuild torch dispatch context skipping `ProxyTorchDispatchMode` instances?

samdow replied:
I think I mostly agree with @ezyang that I don't love this as an idea but I get that copying has some limitations that make this difficult (ideally we would want to copy while passing a new argument to the constructor...). So what about this idea:
(1) delete the inner and ancestor attributes for the copy of every mode in the stack (note: they cannot just be set to None but must be deleted in order to get the mode code to work)
(2) as we're doing this, one by one push the modes back onto the stack

This is still imperfect since we are altering the inner and ancestor elements of the stack but at least the mode mechanism is still the one in charge of the mode ordering instead of having this repeated code

samdow replied:
Also open to thoughts from either of you on this idea

ezyang replied:
I am still advocating for just setting dynamically scoped variables on the context objects themselves. This saves you from (1) having to copy or (2) mutating the inner pointers resulting in weird behavior

samdow replied:
I'm happy with that. And to Ed's other point, I would be happy to have us restrict users updating inner but not sure if there's a clean way to

ezyang reviewed (approved):
I'll probably be editing this in the near future

IvanYashchuk commented:
@pytorchbot merge
